From 43031420261958df0c9b0d336c76c5c6c5d13f15 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Tue, 20 Aug 2024 16:21:56 +0300
Subject: [PATCH] [FSW-7256] fs1: rr configuration for rom recovery

With this patch added support to enable range
register window for all required access zones
with identified initiators of pcie and i2c for
rom recovery.

Change-Id: I896025697531a63a50a168b4a0c7146e60cd31b3
---

diff --git a/arch/fs1/arc/dev_security.c b/arch/fs1/arc/dev_security.c
index be1ba28..ffb3eac 100644
--- a/arch/fs1/arc/dev_security.c
+++ b/arch/fs1/arc/dev_security.c
@@ -303,3 +303,21 @@
 
 	regmap_writel(rr_en_val, initr_data.base_addr + mmRANGE_REG_LBW_RANGE_EN_0 + rr_dist);
 }
+
+void security_rr_recovery_config(bool rr_en)
+{
+	uint8_t i;
+
+	if (rr_en) {
+		for (i = SEC_ACCESS_CTRL_QSPI; i <= SEC_ACCESS_CTRL_PSOC_SEC; i++) {
+			security_rr_lbw_config_en(initiator_data[SECURITY_INITIATOR_PCIE], &lbw_rr_ctx[i]);
+			security_rr_lbw_config_en(initiator_data[SECURITY_INITIATOR_I2C_S_MSTR], &lbw_rr_ctx[i]);
+		}
+
+		/* Incase memory mapped firmware update enable below code */
+		/* security_rr_lbw_config_en(initiator_data[SECURITY_INITIATOR_PCIE], &lbw_rr_ctx[SEC_ACCESS_CTRL_GEN_MEM]); */
+		/* security_rr_lbw_config_en(initiator_data[SECURITY_INITIATOR_I2C_S_MSTR], &lbw_rr_ctx[SEC_ACCESS_CTRL_GEN_MEM]); */
+	} else {
+		/* TODO: Implement if required */
+	}
+}
diff --git a/arch/fs1/arc/dev_security.h b/arch/fs1/arc/dev_security.h
index 4b7d680..55ef2ee 100644
--- a/arch/fs1/arc/dev_security.h
+++ b/arch/fs1/arc/dev_security.h
@@ -107,6 +107,7 @@
 		enum security_img_sts error);
 void security_rr_lbw_config_en(struct security_initiator_ctx initr_data,
 	const struct security_rr_ctx *rr_ip_ctx);
+void security_rr_recovery_config(bool enable);
 
 extern const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN];
 
diff --git a/common/recovery.c b/common/recovery.c
index c19b30f..6929e49 100644
--- a/common/recovery.c
+++ b/common/recovery.c
@@ -15,6 +15,7 @@
 #include <global_conf_regs.h>
 #include <pcie.h>
 #include <common.h>
+#include <dev_security.h>
 
 static void execute_rom_recovery(void)
 {
@@ -25,6 +26,8 @@
 	if (boot_rom_in_master_die())
 		boot_rom_pcie_link_init();
 
+	security_rr_recovery_config(true);
+
 	boot_rom_d2d_link_init();
 
 	boot_rom_status_update(PROGRESS_BOOT_ROM_RECOVERY_END);
